File: wattbuddy-server/services/powerLimitService.js
// 🔔 FEATURE 1: Power-Limit Notification Service
const db = require('../db');

class PowerLimitService {
  // Check power usage against limits and send notifications
  static async checkPowerLimit(userId, currentUsage, dailyLimit) {
    try {
      const thresholds = [50, 75, 90, 100];
      const usagePercentage = (currentUsage / dailyLimit) * 100;

      for (const threshold of thresholds) {
        if (usagePercentage >= threshold) {
          const severity = threshold === 100 ? 'critical' : threshold === 90 ? 'high' : 'warning';
          const title = threshold === 100 ? '🚨 Power Limit Exceeded!' : `⚠️ Warning: ${threshold}% Usage`;
          
          const message = `Your power usage has reached ${usagePercentage.toFixed(1)}% of daily limit (${currentUsage.toFixed(2)}W / ${dailyLimit.toFixed(2)}W)`;

          // Create notification
          await db.query(
            `INSERT INTO notifications 
             (user_id, type, title, message, severity, data)
             VALUES ($1, $2, $3, $4, $5, $6)`,
            [
              userId,
              'power_limit_alert',
              title,
              message,
              severity,
              JSON.stringify({
                current_usage: currentUsage,
                limit: dailyLimit,
                percentage: usagePercentage,
                threshold,
                timestamp: new Date().toISOString()
              })
            ]
          );

          console.log(`✅ Power limit notification sent to user ${userId}: ${threshold}%`);
          return { notificationSent: true, threshold, percentage: usagePercentage };
        }
      }
      return { notificationSent: false };
    } catch (error) {
      console.error('❌ Error checking power limit:', error);
      throw error;
    }
  }

  // Get user's power limit settings
  static async getPowerLimitSettings(userId) {
    try {
      const result = await db.query(
        `SELECT daily_power_limit, alert_threshold 
         FROM user_settings WHERE user_id = $1`,
        [userId]
      );

      return result.rows[0] || { daily_power_limit: 5000, alert_threshold: 0.75 };
    } catch (error) {
      console.error('❌ Error fetching power limit settings:', error);
      return { daily_power_limit: 5000, alert_threshold: 0.75 };
    }
  }

  // Set power limit for user
  static async setPowerLimit(userId, dailyLimit) {
    try {
      const result = await db.query(
        `INSERT INTO user_settings (user_id, daily_power_limit)
         VALUES ($1, $2)
         ON CONFLICT (user_id) 
         DO UPDATE SET daily_power_limit = $2
         RETURNING *`,
        [userId, dailyLimit]
      );

      console.log(`✅ Power limit set for user ${userId}: ${dailyLimit}W`);
      return result.rows[0];
    } catch (error) {
      console.error('❌ Error setting power limit:', error);
      throw error;
    }
  }
}

module.exports = PowerLimitService;

File: lib/services/power_limit_service.dart
import 'package:flutter_local_notifications/flutter_local_notifications.dart';
import 'api_service.dart';

class PowerLimitNotificationService {
  static final FlutterLocalNotificationsPlugin _notificationsPlugin =
      FlutterLocalNotificationsPlugin();

  static Future<void> initialize() async {
    const AndroidInitializationSettings androidSettings =
        AndroidInitializationSettings('@mipmap/ic_launcher');

    const InitializationSettings initSettings =
        InitializationSettings(android: androidSettings);

    await _notificationsPlugin.initialize(initSettings);
  }

  // 🔔 Send power limit notification
  static Future<void> showPowerLimitNotification({
    required String title,
    required String body,
    required double currentPower,
    required double powerLimit,
  }) async {
    const AndroidNotificationDetails androidDetails =
        AndroidNotificationDetails(
      'power_limit_channel',
      'Power Limit Alerts',
      channelDescription: 'Alerts when power usage exceeds limits',
      importance: Importance.high,
      priority: Priority.high,
    );

    const NotificationDetails notificationDetails =
        NotificationDetails(android: androidDetails);

    await _notificationsPlugin.show(
      0,
      title,
      body,
      notificationDetails,
    );
  }

  // Check power limit and notify
  static Future<void> checkAndNotifyPowerLimit(
    String userId,
    double currentPower,
    double powerLimit,
  ) async {
    try {
      final data = await ApiService.post(
        '/power-limit/check',
        {
          'userId': userId,
          'currentUsage': currentPower,
          'dailyLimit': powerLimit,
        },
      );

      if (data['notificationSent'] == true) {
        final percentage = data['percentage'] ?? 0.0;
        final threshold = data['threshold'] ?? 0;

        await showPowerLimitNotification(
          title: threshold == 100
              ? '🚨 Power Limit Exceeded!'
              : '⚠️ Power Usage Warning',
          body:
              'Your power usage is at ${percentage.toStringAsFixed(1)}% of daily limit',
          currentPower: currentPower,
          powerLimit: powerLimit,
        );
      }
    } catch (e) {
      print('❌ Error checking power limit: $e');
    }
  }

  // Set power limit
  static Future<bool> setPowerLimit(
    String userId,
    double dailyLimit,
  ) async {
    try {
      final data = await ApiService.post(
        '/power-limit/set',
        {
          'userId': userId,
          'dailyLimit': dailyLimit,
        },
      );

      return data['success'] == true;
    } catch (e) {
      print('❌ Error setting power limit: $e');
      return false;
    }
  }

  // Get power limit settings
  static Future<Map<String, dynamic>?> getPowerLimitSettings(
    String userId,
  ) async {
    try {
      final data = await ApiService.get(
        '/power-limit/$userId',
      );

      if (data['success'] == true) {
        return {
          'daily_power_limit': data['daily_power_limit'] ?? 5000.0,
          'alert_threshold': data['alert_threshold'] ?? 0.75,
        };
      }
      return null;
    } catch (e) {
      print('❌ Error fetching power limit settings: $e');
      return null;
    }
  }
}
